### PR TITLE
[AIRFLOW-5343] Add pool_pre_ping to SQLAlchemy

### PR DESCRIPTION
**Pull request by Khrol**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-5343
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

SQLalchemy supports connection check while returning it from the pool. There is a need to allow this parameter (pool_pre_ping) while creating the connection to database.
More info here: https://docs.sqlalchemy.org/en/13/core/pooling.html#disconnect-handling-pessimistic

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
It's kinda integration layer with database and not really coverable with unit tests.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by Khrol** [comment]
Tests are red due to flackiness. I've run them two times and they failed differently.

File: airflow/settings.py [hunk 1]
```diff
@@ -173,10 +173,21 @@ def configure_orm(disable_connection_pool=False):
         except conf.AirflowConfigException:
             pool_recycle = 1800
 
+        # Check connection at the start of each connection pool checkout.
+        # Typically, this is a simple statement like “SELECT 1”, but may also make use
+        # of some DBAPI-specific method to test the connection for liveness.
+        # More information here:
+        # https://docs.sqlalchemy.org/en/13/core/pooling.html#disconnect-handling-pessimistic
+        try:
+            pool_pre_ping = conf.getboolean('core', 'SQL_ALCHEMY_POOL_PRE_PING')
```
Comment (ashb) [inline airflow/settings.py]:
```suggestion
            pool_pre_ping = conf.getboolean('core', 'SQL_ALCHEMY_POOL_PRE_PING', fallback=False)
```

And then we don't need the try/except.

**Review by ashb — CHANGES_REQUESTED** [review]
Looks sensible.

What difference in performance/behaviour does enabling this btw? When might a user want it on or not?

**Comment by Khrol** [comment]
> What difference in performance/behaviour does enabling this btw? When might a user want it on or not?

```
The approach adds a small bit of overhead to the connection checkout process, however is otherwise the most simple and reliable approach to completely eliminating database errors due to stale pooled connections.
```

Connection to Airflow backend should not be a bottleneck in Airflow performance. Meanwhile, we are observing flakiness there in our local infrastructure so such overhead might be a good tradeoff.

**Review by ashb — APPROVED** [review]

**Comment by bharathpalaksha** [comment]
@Khrol      setup_event_handlers(engine, reconnect_timeout) is already present at line 198 which is legacy way of handling pessimistic disconnect. pool_pre_ping does the same.

**Comment by Khrol** [comment]
@bharathpalaksha, that's interesting. It means we are observing something different on our local env.

**Comment by Khrol** [comment]
Also does it mean we should remove legacy one and set `sql_alchemy_pool_pre_ping` to default as True?

**Comment by ashb** [comment]
That possibly might make sense, yeah. Can you check the version of SQLA that added pool_pre_ping and check our version is >= to that. (A new PR against the same Jira ticket number please)

**Comment by Khrol** [comment]
@ashb Yes, we are good to go. I'll do a new PR.

https://github.com/apache/airflow/blob/master/setup.py#L357 `'sqlalchemy~=1.3'`

`pool_pre_ping` was added in 1.2

```
New in version 1.2: Added “pre-ping” capability to the Pool class.
```

**Comment by Khrol** [comment]
https://github.com/apache/airflow/pull/6034 is created.

**Pull request merged**